panel/lxqtpanel.cpp

```diff
@@ -660,17 +684,47 @@ void LXQtPanel::setPanelGeometry(bool animate)
         }
     }
 
+    if (!mHidden || !mGeometry.isValid()) mGeometry = rect;
     if(mLayerWindow)
     {
+        // NOTE: On Wayland, QVariantAnimation is used to set appropriate negative margins.
         mLayerWindow->setAnchors(anchors);
-        mLayerWindow->setExclusiveEdge(edge);
-
-        // Make LayerShell apply changes immediatly
-        windowHandle()->requestUpdate();
+        setFixedSize(rect.size());
+        if (animate)
+        {
+            if (mWAnimation == nullptr)
+            {
+                mWAnimation = new QVariantAnimation(this);
+                mWAnimation->setEasingCurve(QEasingCurve::Linear);
+                connect(mWAnimation, &QVariantAnimation::finished, this, [this] {
+                    if (mHidden)
+                        setMargins();
+                });
+                connect(mWAnimation, &QVariantAnimation::valueChanged, this,
+                        [this] (const QVariant &value) {
+                    QMargins margins = layerWindowMargins();
+                    QMarginsF m((mWAnimation->endValue().toReal() - value.toReal())
+                                * mLayerWindow->margins().toMarginsF() / 100
+                                + value.toReal() * margins.toMarginsF() / 100);
+                    mLayerWindow->setMargins(m.toMargins());
+                    windowHandle()->requestUpdate();
+                });
+            }
+            mWAnimation->setDuration(mAnimationTime);
+            mWAnimation->setStartValue(static_cast<qreal>(0));
```

Comment:
Possible enhancements, IMO:
- call `setStartValue` & `setEndValue` only once (after the object creation)
- change `endValue` to 1.0 , then we can avoid the "expensive" division by 100 in logic for `valueChanged` signal

```diff
$ git diff
diff --git a/panel/lxqtpanel.cpp b/panel/lxqtpanel.cpp
index d0320673..6620fd4a 100644
--- a/panel/lxqtpanel.cpp
+++ b/panel/lxqtpanel.cpp
@@ -696,6 +696,8 @@ void LXQtPanel::setPanelGeometry(bool animate)
             {
                 mWAnimation = new QVariantAnimation(this);
                 mWAnimation->setEasingCurve(QEasingCurve::Linear);
+                mWAnimation->setStartValue(static_cast<qreal>(0));
+                mWAnimation->setEndValue(static_cast<qreal>(1));
                 connect(mWAnimation, &QVariantAnimation::finished, this, [this] {
                     if (mHidden)
                         setMargins();
@@ -704,15 +706,13 @@ void LXQtPanel::setPanelGeometry(bool animate)
                         [this] (const QVariant &value) {
                     QMargins margins = layerWindowMargins();
                     QMarginsF m((mWAnimation->endValue().toReal() - value.toReal())
-                                * mLayerWindow->margins().toMarginsF() / 100
-                                + value.toReal() * margins.toMarginsF() / 100);
+                                * mLayerWindow->margins().toMarginsF()
+                                + value.toReal() * margins.toMarginsF());
                     mLayerWindow->setMargins(m.toMargins());
                     windowHandle()->requestUpdate();
                 });
             }
             mWAnimation->setDuration(mAnimationTime);
-            mWAnimation->setStartValue(static_cast<qreal>(0));
-            mWAnimation->setEndValue(static_cast<qreal>(100));
             if (!mHidden)
                 setMargins();
             mWAnimation->start();
```

Reply:
Good points.

I wanted to add a commit for another change but will also put your suggestion into it.